plugins/DDG4EDM4hep/edm4hep/Geant4Output2EDM4hep.cpp

```diff
@@ -0,0 +1,566 @@
+    return ;
+  }
+
+  Geant4ParticleMap* pm = context()->event().extension<Geant4ParticleMap>(false);
+
+  edm4hep::MCParticleCollection* mcpc =
+    const_cast<edm4hep::MCParticleCollection*>(
+      &m_store->get<edm4hep::MCParticleCollection>("MCParticles"));
+
+  //-------------------------------------------------------------------
+  if( typeid( Geant4Tracker::Hit ) == coll->type().type  ){
+
+    edm4hep::SimTrackerHitCollection* sthc =
+      const_cast<edm4hep::SimTrackerHitCollection*>(&m_store->get<edm4hep::SimTrackerHitCollection>(colName));
+
+    for(unsigned i=0 ; i < nhits ; ++i){
+      auto sth = sthc->create() ;
+
+      const Geant4Tracker::Hit* hit = coll->hit(i);
+      const Geant4Tracker::Hit::Contribution& t = hit->truth;
+      int trackID = pm->particleID(t.trackID);
+
+      auto mcp = mcpc->at(trackID);
+
+      sth.setCellID( hit->cellID ) ;
+      sth.setEDep(hit->energyDeposit/CLHEP::GeV);
+      sth.setPathLength(hit->length/CLHEP::mm);
+      sth.setTime(hit->truth.time/CLHEP::ns);
+      sth.setMCParticle(mcp);
+      sth.setPosition({hit->position.x()/CLHEP::mm,
+		       hit->position.y()/CLHEP::mm,
+		       hit->position.z()/CLHEP::mm});
+      sth.setMomentum(edm4hep::Vector3f(hit->momentum.x()/CLHEP::GeV,
+					hit->momentum.y()/CLHEP::GeV,
+					hit->momentum.z()/CLHEP::GeV ));
+
+      auto particleIt = pm->particles().find(trackID);
+      if( ( particleIt != pm->particles().end()) ){
+	// if the original track ID of the particle is not the same as the
+	// original track ID of the hit it was produced by an MCParticle that
+	// is no longer stored
+	sth.setProducedBySecondary( (particleIt->second->originalG4ID != t.trackID) );
+      }
+
+    }
+  //-------------------------------------------------------------------
+  }
+  else if( typeid( Geant4Calorimeter::Hit ) == coll->type().type ){
+
+    Geant4Sensitive*       sd      = coll->sensitive();
+    int hit_creation_mode = sd->hitCreationMode();
+
+    edm4hep::SimCalorimeterHitCollection* schc =
+      const_cast<edm4hep::SimCalorimeterHitCollection*>(
+	&m_store->get<edm4hep::SimCalorimeterHitCollection>(colName));
+
+    colName += "Contributions"  ;
+
+    edm4hep::CaloHitContributionCollection* schcc =
+      const_cast<edm4hep::CaloHitContributionCollection*>(
+	&m_store->get<edm4hep::CaloHitContributionCollection>(colName));
+
+
+    for(unsigned i=0 ; i < nhits ; ++i){
+      auto sch = schc->create() ;
+
+      const Geant4Calorimeter::Hit* hit = coll->hit(i);
+
+      sch.setCellID( hit->cellID );
+      sch.setPosition({float(hit->position.x()/mm),
+		       float(hit->position.y()/mm),
+		       float(hit->position.z()/mm)});
+      sch.setEnergy( hit->energyDeposit );
+
+      // now add the individual step contributions
+      for(Geant4HitData::Contributions::const_iterator ci=hit->truth.begin();
+	  ci!=hit->truth.end(); ++ci){
+
+	auto schc = schcc->create();
```

Comment:
This shadows the `schc` from outside the loop. some slightly longer names for sch/schc/schc/schcc should be considered.